# Pangenome: PGGB seqwish
Tags: #pangenome #pggb #seqwish
🏠 [home](README.md)
***
> [! info] Purpose
> Seqwish uses the wfmash all-vs-all alignments to build a pangenome. One graph will be built per chromosome since every genomes was scaffolded at chromosome-scale.

## split wfmash per chromosome

Alignment and fasta files need to be filtered per chromosome to be able to run it. For the alignments, the filter is applied on the query AND the reference. In grapes, there are 19 chromosomes, the loop goes iteratively over each.

```bash
# subset wfmash
for chr in {01..19}; do awk -v chr="chr${chr}" '$1 ~ chr && $6 ~ chr' wfmash/all.on.all.wfmash_s10000p85n1.paf > wfmash/all.on.all.wfmash_s10000p85n1.chr${chr}.paf; done

# subset fasta
for chr in {01..19}; do grep "chr${chr}" seq/all.fasta | sed 's:>::g' > seq/chr${chr}.id; done

for chr in {01..19}; do seqtk subseq seq/all.fasta seq/chr${chr}.id > seq/chr${chr}.fasta; done
```

## seqwish pangenome construction

https://github.com/pangenome/pggb
https://github.com/ekg/seqwish

Seqwish is run per chromosome.

```bash
for chr in {01..19}; do /usr/bin/time -v -o logs/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.timelog seqwish -p wfmash/all.on.all.wfmash_s10000p85n1.chr${chr}.paf -s seq/chr${chr}.fasta -g seqwish/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.gfa -t 48 -k ${kval} -P 2> logs/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.err
```

## statistics
The following steps are optional, they were used to evaluate the raw pangenome.

### summary stats
```bash
for chr in {01..19}; do vg stats -z seqwish/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.gfa > seqwish/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.stats; done
```

### extract fasta per path
Needed to evaluate the size of the different pangenome classes.

```bash
# extract fasta
for chr in {01..19}; do awk 'BEGIN{FS="\t";OFS="\t"} $1=="S"' seqwish/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.gfa | awk 'BEGIN{FS="\t";OFS="\t"}  {gsub("S", ">s", $1); print $1$2"\n"$3}' > fasta/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.fasta; done

# extract len
for chr in {01..19}; do seqkit fx2tab --length --name fasta/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.fasta > fasta/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.fasta.len; done
```

### extract paths 
Paths have to be extracted. They represent an ordered list of nodes that compose an haplotype/genome. After concatenating at genome-level when we calculate the frequencies, we know that 2 means 2 genomes.
The file genome_loc.txt contains the name of the different genomes (at genome-level, not haplotype-level).

```bash
# extract paths with strand
for chr in {01..19}; do awk '$1=="P"' seqwish/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.gfa | cut -f 2-3 > seqwish/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.path; done

# extract names
for chr in {01..19}; do cut -f 1 seqwish/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.path > seqwish/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.id; done

# from the different chr graph, extract paths and concatenate per haplotype
for chr in {01..19}; do tot_len=$(wc -l seqwish/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.id | cut -f 1 -d ' '); for line_nb in `seq 1 ${tot_len}`; do name=$(sed -n "${line_nb}p" seqwish/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.id); echo "sed -n "${line_nb}p" seqwish/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.path | cut -f 2 | tr ',' '\n' | sed 's:+:\\t+:g' | sed 's:-:\\t-:g' | sed 's:^:s:g'> seqwish/${name}.thru.all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.path"; done; done > extract_path.sh

parallel -j 24 :::: extract_path.sh

# concatenate per genome
for chr in {01..19}; do for genome in $(cat genome_loc.txt); do name=$(basename ${genome}); echo "cat seqwish/${name}.hap[1-2].chr${chr}.thru.all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.path | cut -f 1 | sort -u > seqwish/${name}.thru.all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.path"; done; done > path_concat_perGenome.sh

parallel -j 24 :::: path_concat_perGenome.sh
```

### class paths
After concatenating the nodes of the pangenome without removing duplicates, we can calculate their frequency to class them into core (core; n == 9, all the genomes), dispensable (disp; 1 < n < 9), and private (priv; n == 1) genomes. In this work, the total number of genomes was 9.

```bash
# concatenate all genomes paths
for chr in {01..19}; do for genome in $(cat genome_loc.txt); do name=$(basename ${genome}); cat seqwish/${name}.thru.all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.path; done > seqwish/allGenome.thru.all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.path; done

# get freq per node
for chr in {01..19}; do cat seqwish/allGenome.thru.all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.path | sort | uniq -c > seqwish/allGenome.thru.all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.freq; done

# class
# core
for chr in {01..19}; do awk '$1==9' seqwish/allGenome.thru.all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.freq > seqwish/allGenome.thru.all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.core.freq; done

# disp
for chr in {01..19}; do awk '$1 < 9 && $1 > 1' seqwish/allGenome.thru.all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.freq > seqwish/allGenome.thru.all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.disp.freq; done

# priv
for chr in {01..19}; do awk '$1==1' seqwish/allGenome.thru.all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.freq > seqwish/allGenome.thru.all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.priv.freq; done
```

### stats per class checkpoint
```bash
# get fasta len per class
for chr in {01..19}; do for i in $(ls seqwish/allGenome.thru.all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.*.freq); do class=$(basename $i .freq | sed "s:allGenome.thru.all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.::g"); grep -wFf <(sed 's:^.*s:s:g' $i) seqwish/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.fasta.len > seqwish/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.${class}.fasta.len; done; done
```


[wfmash <- Previous Step](0.02_PGGB_wfmash.md) | [home](README.md) | [Next Step -> smoothxg](0.04_PGGB_smoothxg.md)
