# Pangenome: Pan-GWAS
Tags: #pangenome #pggb #vg #gwas
🏠 [home](README.md)
***

> [! info] Purpose
> Perform the GWAS on the variant genotypes.

## prepare the VCFs
### bgzip the final vcf
```bash
/usr/bin/time -v -o logs/${process}.timelog bgzip --threads 96 vcf/all.on.${graph_name}.map.all.vcf 2> logs/${process}.err
```

### tabix the final vcf
```bash
/usr/bin/time -v -o logs/${process}.timelog tabix -p vcf vcf/all.on.${graph_name}.map.all.vcf.gz 2> logs/${process}.tabix.err
```

### filter missmaf per chr

This is just for having lighter VCFs, hard filters will be applied over later.

```bash
for chr in {01..19}; do /usr/bin/time -v -o logs/${process}.chr${chr}.timelog bcftools filter --threads 48 -e 'F_MISSING > 0.25 || MAF <= 0.01' -o vcf/all.on.${graph_name}.chr${chr}.map.all.missmaf.bcf -O u --regions ${ref_name}.chr${chr} vcf/all.on.${graph_name}.map.all.vcf.gz 2> logs/${process}.chr${chr}.err; done
```

### stats
```bash
# stats per chr
for chr in {01..19}; do /usr/bin/time -v -o logs/${process}.chr${chr}.timelog bcftools stats --threads 48 vcf/all.on.${graph_name}.chr${chr}.map.all.missmaf.bcf > vcf/all.on.${graph_name}.chr${chr}.map.all.missmaf.bcf.stats 2> logs/${process}.chr${chr}.err; done

# total
for chr in {01..19}; do grep 'number of records:' vcf/all.on.reconstruct_${ref_name}.all.chr${chr}.map.all.missmaf.bcf.stats | cut -f 4; done >> vcf/all.on.reconstruct_${ref_name}.all.all.map.all.missmaf.bcf.stats

awk '{ sum += $1 } END { print sum }' vcf/all.on.reconstruct_${ref_name}.all.all.map.all.missmaf.bcf.stats
```

### clean SNPs
```bash
for chr in {01..19}; do echo /usr/bin/time -v -o logs/${process}.chr${chr}.timelog bcftools view -v snps --threads 48 -o vcf/all.on.${graph_name}.chr${chr}.map.all.missmaf.snp.bcf -O u vcf/all.on.${graph_name}.chr${chr}.map.all.missmaf.bcf 2> logs/${process}.chr${chr}.err; done
```

### clean SVs
```bash
for chr in {01..19}; do /usr/bin/time -v -o logs/${process}.chr${chr}.timelog bcftools view --exclude-types snps --threads 48 -o vcf/all.on.${graph_name}.chr${chr}.map.all.missmaf.sv.bcf -O u vcf/all.on.${graph_name}.chr${chr}.map.all.missmaf.bcf 2> logs/${process}.chr${chr}.err; done
```

### stats
```bash
# per chr
for chr in {01..19}; do /usr/bin/time -v -o logs/${process}.chr${chr}.timelog bcftools stats --threads 48 vcf/all.on.${graph_name}.chr${chr}.map.all.missmaf.snp.bcf > vcf/all.on.${graph_name}.chr${chr}.map.all.missmaf.snp.bcf.stats 2> logs/${process}.chr${chr}.err; done

# total
for chr in {01..19}; do grep 'number of records:' vcf/all.on.reconstruct_${ref_name}.all.chr${chr}.map.all.missmaf.snp.bcf.stats | cut -f 4; done >> vcf/all.on.reconstruct_${ref_name}.all.all.map.all.missmaf.snp.bcf.stats

awk '{ sum += $1 } END { print sum }' vcf/all.on.reconstruct_${ref_name}.all.all.map.all.missmaf.snp.bcf.stats
```

## pre-process per variant type
### SNP only
#### concatenate
```bash
ls vcf/*map.all.missmaf.snp.bcf > vcf/all.on.reconstruct_${ref_name}.all.all.map.all.missmaf.snp.bcf.list

/usr/bin/time -v -o logs/${process}.timelog bcftools concat --threads 48 --file-list vcf/all.on.reconstruct_${ref_name}.all.all.map.all.missmaf.snp.bcf.list -O v -o vcf/all.on.reconstruct_${ref_name}.all.map.all.missmaf.snp.vcf 2> logs/${process}.err
```

#### stats
```bash
/usr/bin/time -v -o logs/${process}.timelog bcftools stats --threads 48 vcf/all.on.reconstruct_${ref_name}.all.map.all.missmaf.snp.vcf > vcf/all.on.reconstruct_${ref_name}.all.map.all.missmaf.snp.vcf.stats 2> logs/${process}.err
```

#### simplify the file

Remove extra information not required further.

```bash
/usr/bin/time -v -o logs/${process}.timelog bcftools annotate --threads 48 -x ID,FILTER,FORMAT/GL,FORMAT/GQ,FORMAT/GP,FORMAT/XD,FORMAT/MAD -O v -o vcf/all.on.reconstruct_${ref_name}.all.map.all.missmaf.snp.ez.vcf vcf/all.on.reconstruct_${ref_name}.all.map.all.missmaf.snp.vcf 2> logs/${process}.err
```

#### name variant
```bash
/usr/bin/time -v -o logs/${process}.timelog bcftools annotate --threads 48 --set-id +'%CHROM\_%POS' -O v -o vcf/all.on.reconstruct_${ref_name}.all.map.all.missmaf.snp.ez.variantid.vcf vcf/all.on.reconstruct_${ref_name}.all.map.all.missmaf.snp.ez.vcf 2> logs/${process}.err
```

### SV only

Here, SV are considered as any variants that are not SNPs.

#### concatenate
```bash
/usr/bin/time -v -o logs/${process}.timelog bcftools concat --threads 48 --file-list vcf/all.on.reconstruct_${ref_name}.all.all.map.all.missmaf.sv.bcf.list -O v -o vcf/all.on.reconstruct_${ref_name}.all.map.all.missmaf.sv.vcf 2> logs/${process}.err
```

#### stats
```bash
/usr/bin/time -v -o logs/${process}.timelog bcftools stats --threads 48 vcf/all.on.reconstruct_${ref_name}.all.map.all.missmaf.sv.vcf > vcf/all.on.reconstruct_${ref_name}.all.map.all.missmaf.sv.vcf.stats 2> logs/${process}.err
```

#### simplify the file

Remove extra information not required further.

```bash
/usr/bin/time -v -o logs/${process}.timelog bcftools annotate --threads 48 -x ID,FILTER,FORMAT/GL,FORMAT/GQ,FORMAT/GP,FORMAT/XD,FORMAT/MAD -O v -o vcf/all.on.reconstruct_${ref_name}.all.map.all.missmaf.sv.ez.vcf vcf/all.on.reconstruct_${ref_name}.all.map.all.missmaf.sv.vcf 2> logs/${process}.err
```

#### name variant
```bash
/usr/bin/time -v -o logs/${process}.timelog bcftools annotate --threads 48 --set-id +'%CHROM\_%POS' -O v -o vcf/all.on.reconstruct_${ref_name}.all.map.all.missmaf.sv.ez.variantid.vcf vcf/all.on.reconstruct_${ref_name}.all.map.all.missmaf.sv.ez.vcf 2> logs/${process}.err
```

## quality pre-processing

We will only use the SNP file here as an example. Pipeline is the same for SVs.

### QC
```bash
# add missing tags
bcftools +fill-tags ${vcf_name}.vcf -Ou -o ${vcf_name}.filled.bcf -- -t AN,AC,AF,MAF

# extract info
bcftools query -f '%QUAL\t%DP\t%MAF\n' ${vcf_name}.filled.bcf > ${vcf_name}.filled.qc.txt

# simplify the file
bcftools annotate -x 'FORMAT,^INFO/DP,INFO/AN,INFO/AF,INFO/MAF' -O u -o ${vcf_name}.filled.ez.bcf ${vcf_name}.filled.bcf
```

Based on MAF and DP distribution, filter thresholds are established.

### post QC filtering

Here minimum DP is fixed to the number of samples.

```bash
bcftools filter -e 'MAF <= 0.05 || QUAL < 30 || DP < 153 || DP > 5000' -o all.on.reconstruct_${ref_name}.all.map.all.snp.miss0.25maf0.05qual30dp.variantid.vcf -O v ${vcf_name}.filled.ez.bcf
```

## run GWAS
```bash
# convert vcf
plink --vcf ${vcf_name}.vcf --out ${vcf_name} --recode --allow-extra-chr --make-bed | tee ${vcf_name}_convert.log 2> ${vcf_name}_convert.err

# prune for kinship matrix
plink --bfile ${vcf_name} --indep-pairwise 50 5 0.2 --out ${vcf_name}.indepSNP --allow-extra-chr --allow-no-sex | tee ${vcf_name}.plink_prune.log 2> ${vcf_name}.plink_prune.err

# extract
plink --bfile ${vcf_name} --extract ${vcf_name}.indepSNP.prune.in --out ${vcf_name}.pruned --recode --make-bed --allow-extra-chr --allow-no-sex | tee ${vcf_name}.plink_extract.log 2> ${vcf_name}.plink_extract.err

# make sure to have the .fam file ready
# Relatedness Matrix from Genotypes
gemma -bfile ${vcf_name}.pruned -gk 2 -o ${vcf_name}.pruned.gemmagk2 | tee ${vcf_name}.pruned.gemma_gk2.log 2> ${vcf_name}.pruned.gemma_gk2.err

# gwas
gemma -bfile ${vcf_name} -k output/${vcf_name}.pruned.gemmagk2.sXX.txt -lmm 1 -o ${vcf_name}.gemmagk2_res_lmm1 | tee ${vcf_name}.gemmagk2_res_lmm1.log 2> ${vcf_name}.gemmagk2_res_lmm1.err
```

[Variant genotyping <- Previous Step](0.09_variant_genotyping.md) | [home](README.md)

